llvm/lib/Transforms/Utils/LoopIterCountsToAssumptions.cpp

```diff
@@ -0,0 +1,215 @@
+}
+
+/// get AddRec scalar evolution, if it can be found in Op
+const SCEVAddRecExpr *getAddRecSCEV(Value *Op, ScalarEvolution *SE) {
+  if (!SE->isSCEVable(Op->getType()))
+    return nullptr;
+
+  return dyn_cast<SCEVAddRecExpr>(SE->getSCEV(Op));
+}
+
+/// get the SCEV of the loop that describes how the loop IV is modified
+const SCEVAddRecExpr *getLoopCmpIVAddRecExpr(ICmpInst &LoopCmpInstr,
+                                             ScalarEvolution *SE,
+                                             unsigned &IVOperand) {
+  const SCEVAddRecExpr *S0 = getAddRecSCEV(LoopCmpInstr.getOperand(0), SE);
+  const SCEVAddRecExpr *S1 = getAddRecSCEV(LoopCmpInstr.getOperand(1), SE);
+  if (S0 and S1) {
+    LLVM_DEBUG(
+        dbgs() << "LoopIterCountsToAssumptions-Warning: Encountered two ADDRec "
+                  "expressions in the Loop Compare instruction. Currently it "
+                  " is expected that the compare consists of the IV and a "
+                  "static bound, that is loop invariant.");
+    return nullptr;
+  }
+
+  if (S0) {
+    IVOperand = 0;
+    return S0;
+  }
+
+  if (S1) {
+    IVOperand = 1;
+    return S1;
+  }
+
+  return nullptr;
+}
+
+bool needStartValue(const Value &StartValue, const Value &MinIterValue) {
+  auto *I = dyn_cast<Instruction>(&MinIterValue);
+  if (!I)
+    return true;
+
+  for (Value *Op : I->operands()) {
+    if (Op == &StartValue)
+      return false;
+  }
+  return true;
+}
+
+} // namespace
+
+PreservedAnalyses
+LoopIterCountsToAssumptions::run(Loop &CurrentLoop, LoopAnalysisManager &AM,
+                                 LoopStandardAnalysisResults &AR,
+                                 LPMUpdater &U) {
+  ScalarEvolution *SE = &AR.SE;
+  AssumptionCache *AC = &AR.AC;
+  DominatorTree *DT = &AR.DT;
+
+  std::optional<int> RawMinIterationCount = getMinTripCount(&CurrentLoop);
+
+  // dump loop summary
+  LLVM_DEBUG(if (CurrentLoop.getLoopPreheader()) {
+    dbgs() << "Preheader:" << CurrentLoop.getLoopPreheader()->getName() << "\n";
+  });
+  LLVM_DEBUG(dbgs() << "Header:" << CurrentLoop.getHeader()->getName() << "\n");
+
+  if (!RawMinIterationCount.has_value() || RawMinIterationCount.value() <= 0)
+    return PreservedAnalyses::all();
+  int MinIterCount = RawMinIterationCount.value();
+
+  if (!isLoopRotateCandidate(CurrentLoop)) {
+    LLVM_DEBUG(
+        dbgs()
+        << "LoopIterCountsToAssumptions-Warning: Loop "
+        << CurrentLoop.getHeader()->getParent()->getName() << " "
+        << CurrentLoop.getName()
+        << " will not be rotated, therefore aborting adding assumptions.\n");
+    return PreservedAnalyses::all();
+  }
+
+  LLVM_DEBUG(CurrentLoop.getHeader()->getParent()->dump(););
+  LLVM_DEBUG(dbgs() << "Processing Loop Iteration Count Metadata: "
+                    << CurrentLoop.getHeader()->getParent()->getName() << " "
+                    << CurrentLoop.getName() << " (" << MinIterCount << ")\n");
+  auto *ExitBB = CurrentLoop.getExitingBlock();
```

Comment:
I think that this condition is already covered by `isLoopRotateCandidate`. Look:

```
  /// If getExitingBlocks would return exactly one block, return that block.
  /// Otherwise return null.
  BlockT *getExitingBlock() const;
```

Reply:
I switched to that method